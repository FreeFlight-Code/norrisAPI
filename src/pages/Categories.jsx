import React from "react";
import {connect} from 'react-redux';

class Categories extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      categories: []
    };
  }

  componentDidMount() {
    fetch("https://api.chucknorris.io/jokes/categories")
      .then(res => res.json())
      .then(categories => this.setState({ categories: categories }));
  }

  categoriesList() {
    const { categories } = this.state;
    return categories.map((category, i) => {
      return (
        <div
          className="category"
          onClick={e => {
            this.getJoke(category);
          }}
          key={`category-${i}`}
        >
          {category}
        </div>
      );
    });
  }

  getJoke(category) {
    if (category) {
      fetch(`https://api.chucknorris.io/jokes/random?category=${category}`)
        .then(res => res.json())
        .then(joke => {
          this.props.dispatch({
            type: "DISPLAY_MODAL",
            message: joke.value,
            messageType: "info"
          });
        });
    }
  }

  render() {
    return (
      <div className="page categories">
        <div className="categoryList">{this.categoriesList()}</div>
        {this.getJoke()}
      </div>
    );
  }
}

export default connect()(Categories);
