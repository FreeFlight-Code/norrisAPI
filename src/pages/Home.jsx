import React from 'react';
import Form from '../components/Form';

class Home extends React.Component {

	render() { 

		console.log(this.state)
		return (

      <div className="page home">
        Home Page
		<Form/>
      </div>
    );
	}
}
 
export default Home;