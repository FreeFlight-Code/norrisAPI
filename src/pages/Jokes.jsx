
import React, { Component } from "react";

class Jokes extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }
  render() {
    return <div>Jokes Page</div>;
  }
}

export default Jokes;
