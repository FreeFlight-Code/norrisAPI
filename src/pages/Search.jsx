import React, { useState } from "react";
import { handleJokeClick, truncString, getJokesBySearchTerm } from "../js";

//#functional component
function Search () {

  // #hooks #useState 
  const [searchTerm, setSearchTerm] = useState("");
  const [jokes, setJokes] = useState([]);


  function handleinputChange(val){
    setSearchTerm(val)
  }

  async function handleSubmit(){
    const jokesArray = await getJokesBySearchTerm(searchTerm);
      setJokes(jokesArray);
  }

  return (
    //
    <div style={style.searchPage} className="page search">
      <label htmlFor="searchBar">Search Bar</label>
      {/* #styles type 1 - inline */}
      <input
        data-testid="search-input"
        style={{ padding: "10px 15px" }}
        value={searchTerm}
        id="searchBar"
        type="text"
        onChange={(e) => handleinputChange(e.target.value)}
        placeholder="Search Chuck Norris Joke by Key Word"
      ></input>
      <button
        data-testid="submit-input"
        className="blue"
        onClick={(e) => handleSubmit()}
      >
        Submit
      </button>
      <div className="results">
        <h3>Results</h3>
        <SearchResults jokes={jokes} />
      </div>
    </div>
  );
}

export default Search;

//functional component
export function SearchResults({jokes}) {
  if (jokes && jokes.length > 0) {
    return jokes.map((joke, i) => {
      let shortJoke = truncString(joke.value, 50);
      return (
        <div onClick={e => handleJokeClick(joke)} key={`joke-${i}`}>
          {shortJoke}
        </div>
      );
    });
  } else return null;
}

//#style type 2 - style names will be camelCased rather than with '-' dashes (i.e. backgroundColor rather than background-color)
const style = {
  searchPage: {
    padding: "10vh 3vw",
    background: "linear-gradient(#dddcdc 10%, #b3b2b2)"
  }
};