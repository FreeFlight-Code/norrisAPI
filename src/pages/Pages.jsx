import React from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import { connect } from "react-redux";
import Home from "./Home";
import Categories from "./Categories";
import Jokes from "./Jokes";
import Search from "./Search";

class Routes extends React.Component {

  render() {
    const {loggedin} = this.props.user;
    return (
      <Router>
        {loggedin && <Switch>
          <Route component={Categories} path="/categories" />
          <Route component={Jokes} path="/jokes" />
          <Route component={Search} path="/search" />
        </Switch>}
        <Switch>
          <Route component={Home} exact path="/" />
          <Route path="/">
            <div>404 sorry page not found</div>
          </Route>
        </Switch>
      </Router>
    );
  }
}
 

const mapStateToProps = state => {
  return {
    user: state.user
  };
};

export default connect(mapStateToProps)(Routes);