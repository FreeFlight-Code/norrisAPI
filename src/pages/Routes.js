import React from "react";
import {
  Route,
  Switch,
  useLocation
} from "react-router-dom";
import { connect } from "react-redux";
//components
import Categories from './Categories';
import Jokes from './Jokes';
import Search from './Search';
import Home from './Home';

//#functional component
function Routes(props) {
  //#destructuring
  const {loggedin} = props;
  return (
    <>
{ loggedin && //secured paths only
      <Switch>
        <Route exact path="/categories">
          <Categories />
        </Route>
        <Route exact path="/jokes">
          <Jokes />
        </Route>
        <Route exact path="*">
          <Search />
        </Route>
      </Switch>}

{!loggedin && // non-secured paths only
      <Switch> 
        <Route exact path="/">
          <Home />
        </Route>
        <Route path="*">
          <NoMatch />
        </Route>
      </Switch>}
    </>
  );
}

function NoMatch() {
  let location = useLocation();

  return (
    <div className="page">
      <h3>
        No match for <code>{location.pathname}</code>
      </h3>
      <h3>
        Please login to access secure routes or check your url.
      </h3>
    </div>
  );
}

const mapStateToProps = state =>({
  loggedin: state.user.loggedin
})

export default connect(mapStateToProps)(Routes)