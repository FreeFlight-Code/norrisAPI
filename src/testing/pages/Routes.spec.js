import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

import React from 'react';
import {Routes, NoMatch} from '../../pages/Routes';
import {MemoryRouter as Router} from 'react-router-dom';
import { render, screen } from "@testing-library/react";
import {shallow} from 'enzyme';
import renderer from "react-test-renderer";



describe (' Routes Page ', () => {
	describe (' renders secure routes  ', () => {
		it(" renders home/login page ", () => {
			const page = shallow(
				<Router initialEntries={["/"]}>
					<Routes loggedin={true} />
				</Router>
			);
			const tree = renderer.create(page).toJSON();
        	expect(tree).toMatchSnapshot();
		});
	})
	describe (' renders unsecure routes ', () => {
		it(' renders unknown "/abc" page ', () => {
			const page = shallow(
				<Router initialEntries={["/abc"]}>
					<Routes loggedin={false}/>
				</Router>
			);
			const tree = renderer.create(page).toJSON();
        	expect(tree).toMatchSnapshot();
		})
	})

	// Can we test whether routes work?
});
