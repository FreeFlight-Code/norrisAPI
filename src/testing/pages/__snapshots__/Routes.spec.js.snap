// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Routes Page   renders secure routes    renders home/login page  1`] = `
<div
  className="page search"
  style={
    Object {
      "background": "linear-gradient(#dddcdc 10%, #b3b2b2)",
      "padding": "10vh 3vw",
    }
  }
>
  <label
    htmlFor="searchBar"
  >
    Search Bar
  </label>
  <input
    data-testid="search-input"
    id="searchBar"
    onChange={[Function]}
    placeholder="Search Chuck Norris Joke by Key Word"
    style={
      Object {
        "padding": "10px 15px",
      }
    }
    type="text"
    value=""
  />
  <button
    className="blue"
    data-testid="submit-input"
    onClick={[Function]}
  >
    Submit
  </button>
  <div
    className="results"
  >
    <h3>
      Results
    </h3>
  </div>
</div>
`;

exports[` Routes Page   renders unsecure routes   renders unknown "/abc" page  1`] = `
<div
  className="page"
>
  <h3>
    No match for 
    <code>
      /abc
    </code>
  </h3>
  <h3>
    Please login to access secure routes or check your url.
  </h3>
</div>
`;
